tests/test_packages.py

```diff
@@ -81,6 +83,75 @@ def test_simple_setuptools(tmp_path, env, nvcc_version):
     assert extras == {"test": {"dask-cuda==24.4.*,>=0.0.0a0"}}
 
 
+# rapids-build-backend should support projects using setuptools whose setup.py
+# file has 'import' statements depending on some project(s) that need to be extracted
+# from dependencies.yaml at build time
+def test_setuptools_with_imports_in_setup_py_works(
+    tmp_path, isolated_env, examples_dir
+):
+    package_dir = tmp_path / "pkg"
+    shutil.copytree(
+        src=examples_dir / "setuptools-with-imports-in-setup-py", dst=package_dir
+    )
+
+    with patch_nvcc_if_needed(nvcc_version="85"):
+        name, build_requires, requirements, extras = _generate_wheel(
+            env=isolated_env, package_dir=package_dir
+        )
+
+    assert name == "setuptools-with-imports-in-setup-py-cu85"
+    assert {"more-itertools"}.issubset(build_requires)
+    assert requirements == set()
+
+
+def test_setuptools_with_imports_in_setup_py_fails_on_missing_imports(
+    tmp_path, isolated_env, examples_dir, capfd
+):
+    package_dir = tmp_path / "pkg"
+    shutil.copytree(
+        src=examples_dir / "setuptools-with-imports-in-setup-py", dst=package_dir
+    )
+
+    # only the CUDA '85.*' in this example provides required build dependency
+    # 'more-itertools', so it won't be found if using some other matrix.
+    #
+    # This test confirms that rapids-build-backend fails loudly in that case, instead of
+    # silently ignoring it.
+    #
+    # It'd also catch the case where other tests accidentally pass because
+    # 'more-itertools' already existed in the environment where tests run.
+    with patch_nvcc_if_needed(nvcc_version="25"):
```

Comment:
I'm wondering if we should convert `patch_nvcc_if_needed` into a parametrized autouse fixture. Every test in this file uses it. It might be simpler to do that for CUDA 11 and 12. You could construct dependencies.yaml accordingly on a per-test basis to exercise the necessary cases. Making it an autouse fixture would make including spoofed versions like this harder since it would propagate to all tests, though, but I'm not convinced that it's really necessary since you could just make this dependency conditional on a real version, right?

Reply:
I intentionally uses versions that don't exist in the real world (CUDA 25 and 85) to make it clear that this behavior isn't relying on any special "if CUDA 11" or "if CUDA 12" types of logic in `rapids-build-backend`.

For as long as we want to avoid having such logic in this project (and I think we should), testing like that is a nice way to prevent it from accidentally leaking in.

I'm not sold on an autouse fixture being an improvement over the current state, to be honest. I think it'd add complexity and indirection and I don't see the benefit we'd get in return.

Reply:
I suggested the fixture because basically every test in this file requires the patching and it's not very DRY to have to add this CM in each test. Also, the current approach requires a tighter scope than really matters since you could just do it at the top of every test and be fine.

Reply:
Ok thanks for that. Note that the tests aren't all using the exact same setup.

There are a few (not added by this PR) that `pytest.parametrize()` over CUDA versions...

https://github.com/rapidsai/rapids-build-backend/blob/d2202dd19b08f9f99ea047212ad74720844eb13f/tests/test_packages.py#L55-L56

... and then pass that value into the `patch_nvcc_if_needed()` constructor.

https://github.com/rapidsai/rapids-build-backend/blob/d2202dd19b08f9f99ea047212ad74720844eb13f/tests/test_packages.py#L75

It's hard for me to think of a way to make that consistent with the use of an auto-use fixture in a way that'd be any more DRY than what's already being done here. For example, I don't think it's it's possible to reference fixtures in `pytest.parametrize()` blocks ([which is why plugins like pytest-lazy-fixture were created](https://github.com/tvorog/pytest-lazy-fixture)).

I do also feel that it's helpful to use fake CUDA versions for some of these tests, for as long as we intend to keep specific-to-actual-CUDA-versions functionality out of this project (which I hope we'll be able to keep doing).

So altogether... I don't support switching the `nvcc` patching to an auto-use feature for these tests.